phoenix-core/src/main/java/org/apache/phoenix/util/ServerUtil.java

```diff
@@ -118,6 +122,33 @@ else if (cause == null || cause instanceof IOException) {
         }
     }
 
+    /**
+     * Converts a batch operation status into an IOException if the operation failed partially or completely. In cases
+     * of multiple failures, non-retriable errors are prioritized. If any part of the batch operation failed in a way
+     * that isn't retriable, an instance of the corresponding DoNotRetryIOException subtype is returned, even if other
+     * retriable errors are present. If all parts were completed successfully, null will be returned.
+     */
+    public static IOException createIOException(OperationStatus[] batchOperationStatus) {
+        IOException retriableException = null;
+        for (int i = 0; i < batchOperationStatus.length; i++) {
+            OperationStatus status = batchOperationStatus[i];
+            switch (status.getOperationStatusCode()) {
+                case SUCCESS:
+                    break;
+                case BAD_FAMILY:
+                    return new NoSuchColumnFamilyException(status.getExceptionMsg());
+                case SANITY_CHECK_FAILURE:
+                    return new FailedSanityCheckException(status.getExceptionMsg());
+                case STORE_TOO_BUSY:
+                    retriableException = new RegionTooBusyException(status.getExceptionMsg());
+                    break;
+                default:
+                    return new DoNotRetryIOException(status.getExceptionMsg());
```

Comment:
What about OperationStatus case FAILURE? There are other retriable exceptions other than RegionTooBusyException, aren't there?

Reply:
By looking at Hbase's code it does seem that `FAILURE` is not a retriable state, [example](https://github.com/apache/hbase/blob/branch-2.1/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L3765)

Added a retry for `NOT_RUN` (which if I understand correctly is used when mutation was never executed)